### PR TITLE
[DependencyInjection] Fix autowiring tagged arguments from attributes

### LINKED ISSUE #43272: [DependencyInjection] Non-existent "inner" service when decorated and autowired
**Symfony version(s) affected**: > v5.3.4

**Description**

There is a breaking change since DependencyInjection was updated (patch version !). It works fine from v5.0.0 to v5.3.4 but newer versions have a breaking change with decorated services. It seems it was introduced #42815 because of the pass order.

**How to reproduce**

```php
<?php
namespace Test;
require_once 'vendor/autoload.php';

$containerBuilder = new \Symfony\Component\DependencyInjection\ContainerBuilder();

class Foo {}
class Bar extends Foo {}
class A {
    public function __construct(Foo $foo) {}
}

$containerBuilder->register('foo', Foo::class);

$containerBuilder->register('bar', Bar::class)->setDecoratedService('foo');
$containerBuilder->setAlias(Foo::class, 'bar.inner');

$containerBuilder->register(A::class)->setAutowired(true);

$containerBuilder->compile();
```

This code works for versions <= v5.3.4 and is breaking for versions > v5.3.4.

Error: `PHP Fatal error:  Uncaught Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException: You have requested a non-existent service "bar.inner". in /.../vendor/symfony/dependency-injection/ContainerBuilder.php:979`

**Possible Solution**

Change the pass order that was changed in #42815 (but if I follow correctly it was made to fix #42791 which was made to fix #42347 because autowire arguments were changed in #40406).

### PR DESCRIPTION
**Pull request by okhoshi**
| Q             | A
| ------------- | ---
| Branch?       | 5.3
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #43272
| License       | MIT
| Doc PR        | no

Reimplement #40406 with `AttributeAutoconfigurationPass` to avoid the BC following the change in `CompilerPass` ordering in `PassConfig`.

Also revert the various fix made in an attempt to recover the BC introduced by #40406.

~Note: `5.4` branch was needed because `AttributeAutoconfigurationPass` is not handling parameters in 5.3~

To-do:
- [x] Add a test to cover the breakage presented in #43272

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 1]
```diff
@@ -583,6 +587,12 @@ public function load(array $configs, ContainerBuilder $container)
         $container->registerAttributeForAutoconfiguration(AsController::class, static function (ChildDefinition $definition, AsController $attribute): void {
             $definition->addTag('controller.service_arguments');
         });
+        $container->registerAttributeForAutoconfiguration(TaggedIterator::class, static function (ChildDefinition $definition, TaggedIterator  $attribute, \ReflectionParameter $reflector) {
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: As those attribute live in the DI component itself, I'm wondering whether FrameworkBundle should be responsible for wiring them or whether the ContainerBuilder should automatically wire them (which would make them usable even when using the component standalone). What do you think @nicolas-grekas ?

**Comment by nicolas-grekas** [comment]
Thanks for the PR, I think moving attributes resolution from `AutowirePass` to `AttributeAutoconfigurationPass` is a nice approach.

Yet I see three issues that should be fixed to me:
- this targets 5.4 while 5.3 is affected
- this requires enabling autoconfiguration (as seen in patched tests)
- this moves registering the attributes to FrameworkBundle (as seen in patched tests and highlighted by @stof).

I think all 3 issues can be solved by patching `AttributeAutoconfigurationPass` on 5.3 so that it hardcodes a special behavior for those two attributes (they were already hardcoded in `AutowirePass` anyway.). This should allow *not* patching most test cases, which is desired.

**Comment by okhoshi** [comment]
Thanks for the comments !

Is it possible to target 5.3 since the `AttributeAutoconfigurationPass` has drastically changed with 5.4 ? AFAIK, there's no logic to apply attributes from parameters, only on the classes.

When you say that you would hardcode the attributes in the `AttributeAutoconfigurationPass`, do you mean that you would register them in the `$parameterAttributeConfigurators` or that you would hardcode the transformation in `processValue` directly ?
As a personal taste, I would do the former, but it would still require enabling the autoconfiguration. By passing the autoconfiguration would likely require deep changes in `AttributeAutoconfiguration`.

As a side note, I had to change the behavior of `ResolveInstanceofConditionalsPass` to recursively resolve nested Definition (without that, one of the tests was not passing anymore).

**Comment by nicolas-grekas** [comment]
We need to fix this issue in 5.3, and I don't have a better idea.
We should not require autconfiguration to be enabled as that'd be a behavior change (booo scary BC breaks :ghost:).
It'd suggest starting with a copy/paste from 5.4 and then removing everything that is not strictly needed to fix this issue, while keeping the common parts as near as possible to ease merging up.

> I had to change the behavior of ResolveInstanceofConditionalsPass to recursively resolve nested Definition

Yes, I didn't full grasp this part but I'll review more carefully after the rest is fixed :)

**Comment by okhoshi** [comment]
@nicolas-grekas I pushed a rework of the branch. It works with the IntegrationTest, but it breaks some tests on the side, mainly because the `AttributeAutoconfigurationPass` needs to run every time now (I spotted errors in the tests because getConstructor was called on a Definition for which the factory was not existing, not sure how I should handle that ?)

I'll fix the other tests tomorrow.

Also, I noticed that #43386 has been merged on 5.4 only and it's modifying the tags, so conflicts will arise :(.

File: src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php [hunk 2]
```diff
@@ -20,36 +24,84 @@
...
+            || !($classReflector = $this->container->getReflectionClass($value->getClass(), false))
         ) {
             return parent::processValue($value, $isRoot);
         }
 
-        $autoconfiguredAttributes = $this->container->getAutoconfiguredAttributes();
+        $classAttributeConfigurators = $value->isAutoconfigured() ? $this->classAttributeConfigurators : [];
+        $parameterAttributeConfigurators = array_merge(
+            $value->isAutoconfigured() ? $this->parameterAttributeConfigurators : [],
+            [
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php]: this part should be merged only if the definition is autowired

Comment (okhoshi) [reply src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php]: Fixed

**Review by nicolas-grekas — APPROVED** [review]
While reviewing more carefully, I found that patching `AttributeAutoconfigurationPass` would not work well with autowiring of `@required` methods, or with named arguments, etc.

~Instead, I pushed a new pass that is dedicated to autowiring tagged arguments from attributes. This pass runs after `AutowireRequired*Pass` and before `ServiceLocatorTagPass`, thus fixing all concerns.~

**Comment by nicolas-grekas** [comment]
I figured out an even simpler patch, PR updated.

**Comment by nicolas-grekas** [comment]
Thank you @Okhoshi.

**Pull request merged**